api/envoy/extensions/filters/http/transform/v3/transform.proto

```diff
@@ -0,0 +1,83 @@
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.transform.v3";
+option java_outer_classname = "TransformProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/transform/v3;transformv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Transform filter configuration]
+// Transform filter :ref:`configuration overview <config_http_filters_transform>` to perform
+// HTTP header and body transformations.
+// [#extension: envoy.filters.http.transform]
+
+// Configuration for the transform filter. The request and response transformations are
+// independent and could be configured separately.
+// Only JSON body transformation is supported for now.
+message TransformConfig {
+  // Configuration for transforming request. If set then the request will be buffered
+  // until the entire body is received and transformed for JSON request bodies.
+  Transformation request_transformation = 1;
+
+  // Configuration for transforming response. If set then the response will be buffered
+  // until the entire body is received and transformed for JSON response bodies.
+  Transformation response_transformation = 2;
+
+  // If true and the request headers are transformed, Envoy will re-evaluate the target
+  // cluster in the same route. Please ensure the cluster specifier in the route supports
+  // dynamic evaluation or this flag will have no effect, e.g.
+  // :ref:`matcher cluster specifier
+  // <envoy_v3_api_msg_extensions.router.cluster_specifiers.matcher.v3.MatcherClusterSpecifier>`.
+  //
+  // Only one of ``clear_cluster_cache`` and ``clear_route_cache`` can be true.
+  bool clear_cluster_cache = 3;
+
+  // If true and the request headers are transformed, Envoy will clear the route cache for
+  // the current request and force re-evaluation of the route. This has performance penalty and
+  // should only be used when the route match criteria depends on the transformed headers.
+  //
+  // Only one of ``clear_cluster_cache`` and ``clear_route_cache`` can be true.
+  bool clear_route_cache = 4;
+}
+
+message Transformation {
+  // The header mutations to perform.
+  // The :ref:`substitution format specifier <config_access_log_format>` could be applied here.
+  // And except the commonly used format specifiers, there are some additional format specifiers
+  // provided by the transform filter:
+  //
+  // * ``%REQUEST_BODY(KEY*)%``: the request body. And ``Key`` KEY is an optional
+  //   lookup key in the namespace with the option of specifying nested keys separated by ':'.
+  // * ``%RESPONSE_BODY(KEY*)%``: the response body. And ``Key`` KEY is an optional
+  //   lookup key in the namespace with the option of specifying nested keys separated by ':'.
+  repeated config.common.mutation_rules.v3.HeaderMutation headers_mutations = 1;
```

Comment:
Will the headers be blocked until the body is buffered even if `headers_mutations` isn't configured?

Reply:
I understand that this is an implementation issue, but I think it should be spelled clearly in a note above (example: buffering of the entire request headers and body will always happen on a request even if only headers are transformed).